<!-- -->
<launch>

  <arg name="time_synchro_threshold" default="0.02" />
  
  <arg name="use_every_scan" default="true" />
  <arg name="frame_dist" default="0.3" /> <!-- only work when use_every_scan is false -->
  <!-- filter -->
  <arg name="scan_calibration_angle" default="0" />   <!-- 10.29 bag 90 -->
  <arg name="min_scan_range" default="5" />
  <arg name="max_scan_range" default="60" />
  <arg name="voxel_filter_size" default="1" />        <!-- ndt 1.0   icp 0.1~0.2-->
  
  <!-- regis-->
  <arg name="method_type" default="0" />              <!-- 0-PCL_GENERIC 1-ICP -->
  <!-- regis-ndt-config -->
  <arg name="ndt_trans_epsilon" default="0.01" />
  <arg name="ndt_step_size" default="0.1" />
  <arg name="ndt_resolution" default="1.0" />
  <arg name="ndt_maxiterations" default="30" />
  <!-- regis-icp-config -->
  <arg name="segment_config_file" default="/autoware/workspace/data/params/lidar_odometry_config.yaml" />
  <arg name="icp_config_file" default="/autoware/workspace/data/params/libpointmatcher_config.p2plane.yaml" />
    <!-- <arg name="icp_config_file" default="/autoware/workspace/data/params/libpointmatcher_config.p2p.yaml" /> -->
  
  <!-- ceres-optimizer -->
  <arg name="optimize_num_every_scans" default="30" />
  <arg name="optimize_iter_num" default="60" />
  <arg name="optimize_var_anchor" default="0.8" />
  <arg name="optimize_var_odom_t" default="0.7" />
  <arg name="optimize_var_odom_q" default="0.7" />
  <!-- <arg name="optimize_var_anchor" default="0.8" />
  <arg name="optimize_var_odom_t" default="0.2" />
  <arg name="optimize_var_odom_q" default="0.02" /> -->
  <arg name="map_init_position_file" default="/autoware/workspace/data/map/new_map_position.txt" />
  <arg name="map_cloud_update_interval" default="5" />   <!-- unit/second -->
  <arg name="keyframe_delta_trans" default="0.3" />
  <!--config over-->
  
  <!-- rosrun lidar_localizer ndt_mapping  -->
  <!-- <node pkg="lidar_localizer" type="queue_counter" name="queue_counter" output="screen"/> -->
  <!-- <node pkg="rviz" type="rviz" name="ceres_mapping_rviz" required="true" 
      args = "-d /autoware/workspace/data/lidar_mapping_ceres.rviz" /> -->
  
  <!-- nmea2tfpose -->
  <!-- <include file="$(find gnss_localizer)/launch/nmea2tfpose.launch"/> -->

  
  <node pkg="lidar_localizer" type="lidar_mapping_ceres" name="lidar_mapping_ceres" output="screen"   >
    <!-- time_config -->
    <param name="time_synchro_threshold" value="$(arg time_synchro_threshold)" />
    <param name="use_every_scan" value="$(arg use_every_scan)" />
    <param name="frame_dist" value="$(arg frame_dist)" />
    <!-- proprecess filter_params -->
    <param name="scan_calibration_angle" value="$(arg scan_calibration_angle)" />
    <param name="min_scan_range" value="$(arg min_scan_range)" />
    <param name="max_scan_range" value="$(arg max_scan_range)" />
    <param name="voxel_filter_size" value="$(arg voxel_filter_size)" />
    <!-- registration params -->
    <param name="method_type" value="$(arg method_type)" />

    <!-- ndt-config -->
    <param name="ndt_trans_epsilon" value="$(arg ndt_trans_epsilon)" />
    <param name="ndt_step_size" value="$(arg ndt_step_size)" />
    <param name="ndt_resolution" value="$(arg ndt_resolution)" />
    <param name="ndt_maxiterations" value="$(arg ndt_maxiterations)" />
    <!-- icp-config -->
    <param name="segment_config_file" value="$(arg segment_config_file)" />
    <param name="icp_config_file" value="$(arg icp_config_file)" />
    
    <!-- back_end optimizer params -->
    <param name="optimize_num_every_scans" value="$(arg optimize_num_every_scans)" />
    <param name="optimize_iter_num" value="$(arg optimize_iter_num)" />
    <param name="optimize_var_anchor" value="$(arg optimize_var_anchor)" />
    <param name="optimize_var_odom_t" value="$(arg optimize_var_odom_t)" />
    <param name="optimize_var_odom_q" value="$(arg optimize_var_odom_q)" />
    <!-- mapping -->
    <param name="map_init_position_file" value="$(arg map_init_position_file)" />
    <param name="map_cloud_update_interval" value="$(arg map_cloud_update_interval)" />
    <param name="keyframe_delta_trans" value="$(arg keyframe_delta_trans)" />
  </node>

</launch>
